Code/Engine/RendererCore/Meshes/Implementation/DynamicMeshBufferResource.cpp

```diff
@@ -0,0 +1,193 @@
+
+  {
+    ezVertexStreamInfo si;
+    si.m_uiOffset = 0;
+    si.m_Format = ezGALResourceFormat::XYZFloat;
+    si.m_Semantic = ezGALVertexAttributeSemantic::Position;
+    si.m_uiElementSize = sizeof(ezVec3);
+    m_VertexDeclaration.m_VertexStreams.PushBack(si);
+
+    si.m_uiOffset += si.m_uiElementSize;
+    si.m_Format = ezGALResourceFormat::XYFloat;
+    si.m_Semantic = ezGALVertexAttributeSemantic::TexCoord0;
+    si.m_uiElementSize = sizeof(ezVec2);
+    m_VertexDeclaration.m_VertexStreams.PushBack(si);
+
+    si.m_uiOffset += si.m_uiElementSize;
+    si.m_Format = ezGALResourceFormat::XYZFloat;
+    si.m_Semantic = ezGALVertexAttributeSemantic::Normal;
+    si.m_uiElementSize = sizeof(ezVec3);
+    m_VertexDeclaration.m_VertexStreams.PushBack(si);
+
+    si.m_uiOffset += si.m_uiElementSize;
+    si.m_Format = ezGALResourceFormat::XYZFloat;
+    si.m_Semantic = ezGALVertexAttributeSemantic::Tangent;
+    si.m_uiElementSize = sizeof(ezVec3);
+    m_VertexDeclaration.m_VertexStreams.PushBack(si);
+
+    m_VertexDeclaration.ComputeHash();
+  }
+
+  ezGALDevice* pDevice = ezGALDevice::GetDefaultDevice();
+
+  m_hVertexBuffer = pDevice->CreateVertexBuffer(sizeof(ezDynamicMeshVertex), m_Descriptor.m_uiNumVertices /* no initial data -> mutable */);
+
+  ezStringBuilder sName;
+  sName.Format("{0} - Dynamic Vertex Buffer", GetResourceDescription());
+  pDevice->GetBuffer(m_hVertexBuffer)->SetDebugName(sName);
+
+  if (m_Descriptor.m_uiNumIndices32 > 0)
+  {
+    m_Index32Data.SetCountUninitialized(m_Descriptor.m_uiNumIndices32);
+
+    m_hIndexBuffer = pDevice->CreateIndexBuffer(ezGALIndexType::UInt, m_Descriptor.m_uiNumIndices32 /* no initial data -> mutable */);
+
+    sName.Format("{0} - Dynamic Index32 Buffer", GetResourceDescription());
+    pDevice->GetBuffer(m_hIndexBuffer)->SetDebugName(sName);
+  }
+  else if (m_Descriptor.m_uiNumIndices16 > 0)
+  {
+    m_Index16Data.SetCountUninitialized(m_Descriptor.m_uiNumIndices16);
+
+    m_hIndexBuffer = pDevice->CreateIndexBuffer(ezGALIndexType::UShort, m_Descriptor.m_uiNumIndices16 /* no initial data -> mutable */);
+
+    sName.Format("{0} - Dynamic Index16 Buffer", GetResourceDescription());
+    pDevice->GetBuffer(m_hIndexBuffer)->SetDebugName(sName);
+  }
+
+  // we only know the memory usage here, so we write it back to the internal variable directly and then read it in UpdateMemoryUsage() again
+  ModifyMemoryUsage().m_uiMemoryGPU = m_VertexData.GetHeapMemoryUsage() + m_Index32Data.GetHeapMemoryUsage();
+
+  ezResourceLoadDesc res;
+  res.m_uiQualityLevelsDiscardable = 0;
+  res.m_uiQualityLevelsLoadable = 0;
+  res.m_State = ezResourceState::Loaded;
+
+  return res;
+}
+
+void ezDynamicMeshBufferResource::UpdateGpuBuffer(ezGALCommandEncoder* pGALCommandEncoder, ezUInt32 uiFirstVertex, ezUInt32 uiNumVertices, ezUInt32 uiFirstIndex, ezUInt32 uiNumIndices)
+{
+  // TODO: can ezGALUpdateMode::Discard be used, if only a subset is updated (without destroying existing data) ?
```

Comment:
No discard literally means "discard existing data". If you update only a subset but don't want to destroy existing data you need to use CopyToTempStorage